homeassistant/components/remootio/cover.py

```diff
@@ -0,0 +1,184 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_ENTITY_ID, ATTR_NAME, CONF_DEVICE_CLASS
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import ATTR_SERIAL_NUMBER, CONF_SERIAL_NUMBER, DOMAIN, REMOOTIO_CLIENT
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up an ``RemootioCover`` entity based on the given configuration entry."""
+
+    _LOGGER.debug("config_entry [%s]", config_entry.as_dict())
+    _LOGGER.debug(
+        "hass.data[%s][%s] [%s]]",
+        DOMAIN,
+        config_entry.entry_id,
+        hass.data[DOMAIN][config_entry.entry_id],
+    )
+
+    serial_number: str = config_entry.data[CONF_SERIAL_NUMBER]
+    device_class: str = config_entry.data[CONF_DEVICE_CLASS]
+    remootio_client: RemootioClient = hass.data[DOMAIN][config_entry.entry_id][
+        REMOOTIO_CLIENT
+    ]
+
+    async_add_entities(
+        [
+            RemootioCover(
+                serial_number, config_entry.title, device_class, remootio_client
+            )
+        ]
+    )
+
+
+class RemootioCover(cover.CoverEntity):
+    """Cover entity which represents an Remootio device controlled garage door or gate."""
+
+    _remootio_client: RemootioClient
+
+    def __init__(
+        self,
+        unique_id: str,
+        name: str,
+        device_class: str,
+        remootio_client: RemootioClient,
+    ) -> None:
+        """Initialize this cover entity."""
+        super().__init__()
+        self._attr_name = name
+        self._attr_unique_id = unique_id
+        self._attr_device_class = device_class
+        self._remootio_client = remootio_client
+        self._attr_device_info = DeviceInfo(
+            name="Remootio",
+            manufacturer="Assemblabs Ltd",
+            suggested_area="The first end-to-end encrypted Wi-Fi and Bluetooth enabled smart remote controller, "
+            "that lets you control and monitor your gates and garage doors using your smartphone.",
+        )
+        self._attr_should_poll = False
+        self._attr_supported_features = cover.SUPPORT_OPEN | cover.SUPPORT_CLOSE
+
+    async def async_added_to_hass(self) -> None:
+        """Register listeners to the used Remootio client to be notified on state changes and events."""
+        await self._remootio_client.add_state_change_listener(
+            RemootioCoverStateChangeListener(self)
+        )
+
+        await self._remootio_client.add_event_listener(RemootioCoverEventListener(self))
+
+        self.async_schedule_update_ha_state(force_refresh=True)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Terminate the used Remootio client."""
+        await self._remootio_client.terminate()
```

Comment:
Client should not be terminated here. That should be done in unload of integration.

Reply:
Where should be this be done then? According to the [documentation](https://developers.home-assistant.io/docs/config_entries_index?_highlight=async_remove#unloading-entries) it must be done here.

Reply:
It should terminate when integration unload. Not when entity unload. That is here: async_unload_entry(hass

Like mentioned earlier with the comment on update coordinator, an integration can (and usually should) have multiple entities, all sharing a single coordinated update client.

Reply:
Yes, `async_will_remove_from_hass` will be invoked when an entity is disabled in HASS, but when it is enabled again then `async_unload_entry` and `async_setup_entry` are invoked in this order based on my exploratory tests.

Reply:
No they will not. Only if you unload and reload the integration.

Reply:
If I deactivate the entity over the GUI, following will be logged:

```
2022-06-10 21:07:24 DEBUG (MainThread) [homeassistant.components.remootio.cover] Doing async_will_remove_from_hass
2022-06-10 21:07:24 INFO (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Terminating this client...
2022-06-10 21:07:24 INFO (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Receiving and handling of messages by this client will be now stopped because of cancelling the task.
2022-06-10 21:07:24 INFO (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Sending PINGs by this client will be now stopped because of cancelling the task.
2022-06-10 21:07:24 DEBUG (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Invoking event listeners... Event [Event(source=<EventSource.CLIENT: 'client'>, type=<EventType.TASKS_STOPPED: 'tasks_stopped'>, key=None)]
2022-06-10 21:07:24 DEBUG (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Task to disconnect from the device isn't started yet. Waiting as long as it is started.
2022-06-10 21:07:25 INFO (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Disconnecting from the device...
```

If I activate the deactivated entity over the GUI, following will be logged regardless whether HASS was restarted in the meantime or not:

```
2022-06-10 21:08:31 INFO (MainThread) [homeassistant.config_entries] Reloading configuration entries because disabled_by changed in entity registry: 
2022-06-10 21:08:31 DEBUG (MainThread) [homeassistant.components.remootio] Doing async_unload_entry. entry [{'entry_id': '9e109cee95ad7e82bb95389954a0d819', 'version': 1, 'domain': 'remootio', 'title': 'Remootio Device (IP: 10.49.56.53, S/N: 2462abe67024duriwsrs)', 'data': {'device_class': 'gate', 'host': '10.49.56.53', 'secret__api_secret_key': '682F864C13345DF05C409D82431F8697C6AC46556699CE41BFB7C827F20AE3F2', 'secret__api_auth_key': '38EE51DFC7D71F03E0E0D448EC282304FA241E24CD6815076F1FB7B827077C6C', 'serial_number': '2462abe67024duriwsrs'}, 'options': {}, 'pref_disable_new_entities': False, 'pref_disable_polling': False, 'source': 'user', 'unique_id': '2462abe67024duriwsrs', 'disabled_by': None}] hass.data[remootio][9e109cee95ad7e82bb95389954a0d819] [{'remootio_client': RemootioClient{host:10.49.56.53,serial_number:2462abe67024duriwsrs,api_version:3,state:State.CLOSED}}]
2022-06-10 21:08:31 DEBUG (MainThread) [homeassistant.components.remootio] Doing async_setup_entry. entry [{'entry_id': '9e109cee95ad7e82bb95389954a0d819', 'version': 1, 'domain': 'remootio', 'title': 'Remootio Device (IP: 10.49.56.53, S/N: 2462abe67024duriwsrs)', 'data': {'device_class': 'gate', 'host': '10.49.56.53', 'secret__api_secret_key': '682F864C13345DF05C409D82431F8697C6AC46556699CE41BFB7C827F20AE3F2', 'secret__api_auth_key': '38EE51DFC7D71F03E0E0D448EC282304FA241E24CD6815076F1FB7B827077C6C', 'serial_number': '2462abe67024duriwsrs'}, 'options': {}, 'pref_disable_new_entities': False, 'pref_disable_polling': False, 'source': 'user', 'unique_id': '2462abe67024duriwsrs', 'disabled_by': None}]
2022-06-10 21:08:31 INFO (MainThread) [homeassistant.components.remootio.aioremootio.remootioclient] Initializing this client...
2022-06-10 21:08:31 INFO (MainThread) [homeassistant.components.cover] Setting up cover.remootio
2022-06-10 21:08:31 DEBUG (MainThread) [homeassistant.components.remootio.cover] Doing async_setup_entry. config_entry [{'entry_id': '9e109cee95ad7e82bb95389954a0d819', 'version': 1, 'domain': 'remootio', 'title': 'Remootio Device (IP: 10.49.56.53, S/N: 2462abe67024duriwsrs)', 'data': {'device_class': 'gate', 'host': '10.49.56.53', 'secret__api_secret_key': '682F864C13345DF05C409D82431F8697C6AC46556699CE41BFB7C827F20AE3F2', 'secret__api_auth_key': '38EE51DFC7D71F03E0E0D448EC282304FA241E24CD6815076F1FB7B827077C6C', 'serial_number': '2462abe67024duriwsrs'}, 'options': {}, 'pref_disable_new_entities': False, 'pref_disable_polling': False, 'source': 'user', 'unique_id': '2462abe67024duriwsrs', 'disabled_by': None}] hass.data[remootio][9e109cee95ad7e82bb95389954a0d819] [{'remootio_client': RemootioClient{host:10.49.56.53,serial_number:N/A,api_version:N/A,state:State.UNKNOWN}}]
2022-06-10 21:08:31 DEBUG (MainThread) [homeassistant.components.remootio.cover] Doing async_added_to_hass
```

Reply:
Yes home assistant will often reload the integration when disabled entities are enabled. Because most doesnt support any other way of adding them. Doesnt change the fact that if you have two entities, will remove will be called on one of the entities but not the other. The other is expected to continue to function.

That you only have one entity now doesnt does not mean we allow wrong design hindering additions later.

Reply:
> Yes home assistant will often reload the integration when disabled entities are enabled. Because most doesnt support any other way of adding them. Doesnt change the fact that if you have two entities, will remove will be called on one of the entities but not the other. The other is expected to continue to function.
> 
> That you only have one entity now doesnt does not mean we allow wrong design hindering additions later.

Please let me understand. Using the config flow for each controlled device (in my case each Remootio) an config entry will be created. During the startup of Home Assistant for each config entry
`homeassistant.components.remootio.async_setup_entry(homeassistant.core.HomeAssistant, homeassistant.config_entries.ConfigEntry)` will be invoked which creates based on the passed config entry an instance of `aioremootio.RemootioClient` and invokes indirectly `homeassistant.components.remootio.cover.async_setup_entry(homeassistant.core.HomeAssistant, homeassistant.config_entries.ConfigEntry, homeassistant.helpers.entity_platform.AddEntitiesCallback)` for the current config entry. The latter method creates an instance of `homeassistant.components.remootio.cover.RemootioCover` which uses the previously created `aioremootio.RemootioClient` instance to communicate with the by the passed config entry represented Remootio. Do I understand the process well? 

Am I right about that the method `async_added_to_hass()` of the created `RemootioCover` instance will be only invoked by Home Assistant during the startup if the entity is activated, furthermore when Home Assistant is started and an deactivated entity is activated?

Am I right about that the method `async_will_remove_from_hass()` of the created `RemootioCover` instance will be invoked by Home Assistant for an active entity if it is deactivated?

When will be `homeassistant.components.remootio.async_unload_entry(...)` invoked?

Thank you!